### PR TITLE
Allow eliding name in pulumi.Config .ctor

### LINKED ISSUE #1581: pulumi.Config('projectName'); Why projectName?
Why does Pulumi need to know the project? Does it not inherently know the Project from the context it's executing in? 

Especially since the configuration values are not defined by the project, but rather by seemingly just existing on the Stack?

If anything, I'd expect to have to input the Stack name. But then, I'm even more confused after skimming through: https://github.com/pulumi/pulumi/issues/1556

### PR DESCRIPTION
**Pull request by ellismg**
When this argument is not provided, we'll default to the value of
pulumi.getProject(). This is what you want for application level code
anyway and it matches the CLI behavior where if you don't qualify a
key with a package we use the name of the current project.

Fixes #1581

### CODE REVIEW

**Comment by ellismg** [comment]
I didn't do this for golang, since it felt a little weird there, but if we want "" to mean "use the current project" I'm happy to touch that as well.

**Comment by ellismg** [comment]
@CyrusNajmabadi I believe the tests that are failing are failing because we now have to capture the some additional stuff from the metadata module.  Is there a way to update these baselines (and can you take a look at this to tell me what I'm doing would work here?  I believe that it should)

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I believe the tests that are failing are failing because we now have to capture the some additional stuff from the metadata module. Is there a way to update these baselines (and can you take a look at this to tell me what I'm doing would work here? I believe that it should)

I need clarity on this.  Does project.getProject even work at cloud-run-time?

**Comment by CyrusNajmabadi** [comment]
from a cursory look at the code, this depends on code like:

```ts
    return {
        project: process.env["PULUMI_NODEJS_PROJECT"],
```

So this would not work at runtime.

For this to work, we would have to push taht computation up to a higher location such that this code executed at *deployment time*.  We'd then capture the resultant values, and that would ensure that the values could be used at cloud-run-time.

**Comment by lukehoban** [comment]
> For this to work, we would have to push taht computation up to a higher location such that this code executed at deployment time. We'd then capture the resultant values, and that would ensure that the values could be used at cloud-run-time.

Yep - that seems like a reasonable approach to support this.

**Comment by ellismg** [comment]
@CyrusNajmabadi @swgillespie Can you take another look here?  After the previous attempt and inspecting code, I realized that our lazy initialization strategy of both config and settings was fragile in the fact of multiple loaded copies of `@pulumi/pulumi`.  In 5eb78af I removed the lazy initialization (such that both settings and config are "locked in" when their respective modules are loaded).  This requires a little crazy dance in `run.ts` to be able to set some environment variables (the only stable store of settings we have when multiple copies of `@pulumi/pulumi` could be loaded).  The second change is as before, but I also did the small amount of work to put golang on the same plan.

File: sdk/nodejs/config.ts [hunk 1]
```diff
@@ -25,11 +26,16 @@ import { getConfig } from "./runtime";
...
-     * name is the configuration bag's logical name and uniquely identifies it.
+     * name is the configuration bag's logical name and uniquely identifies it.  The default is the name of the current
+     * project.
      */
     public readonly name: string;
 
-    constructor(name: string) {
+    constructor(name?: string) {
+        if (name === undefined) {
+            name = getProject();
```
[the remark this answers is not in the document]
Comment (ellismg) [reply sdk/nodejs/config.ts]: Yes, it is.

**Pull request merged**